src/transforms/transaction.rs

```diff
@@ -0,0 +1,812 @@
+        MergeStrategy::Discard => Ok(Box::new(DiscardMerger::new(v))),
+    }
+}
+
+//------------------------------------------------------------------------------
+
+struct TransactionState {
+    fields: HashMap<String, Box<dyn TransactionValueMerger>>,
+    stale_since: Instant,
+}
+
+impl TransactionState {
+    fn new(e: LogEvent, strategies: &IndexMap<String, MergeStrategy>) -> Self {
+        Self {
+            stale_since: Instant::now(),
+            // TODO: all_fields alternative that consumes
+            fields: e
+                .all_fields()
+                .filter_map(|(k, v)| {
+                    if let Some(strat) = strategies.get(&k) {
+                        match get_value_merger(v.clone(), strat) {
+                            Ok(m) => Some((k, m)),
+                            Err(err) => {
+                                warn!("failed to create merger for field '{}': {}", k, err);
+                                None
+                            }
+                        }
+                    } else {
+                        Some((k, v.clone().into()))
+                    }
+                })
+                .collect(),
+        }
+    }
+
+    fn add_event(&mut self, e: LogEvent, strategies: &IndexMap<String, MergeStrategy>) {
+        for (k, v) in e.all_fields() {
+            let strategy = strategies.get(&k);
+            match self.fields.entry(k) {
+                hash_map::Entry::Vacant(entry) => {
+                    if let Some(strat) = strategy {
+                        match get_value_merger(v.clone(), strat) {
+                            Ok(m) => {
+                                entry.insert(m);
+                            }
+                            Err(err) => {
+                                warn!("failed to merge value: {}", err);
+                            }
+                        }
+                    } else {
+                        entry.insert(v.clone().into());
+                    }
+                }
+                hash_map::Entry::Occupied(mut entry) => {
+                    if let Err(err) = entry.get_mut().add(v.clone()) {
+                        warn!("failed to merge value: {}", err);
+                    }
+                }
+            }
+        }
+        self.stale_since = Instant::now();
+    }
+
+    fn flush(mut self) -> LogEvent {
+        let mut event = Event::new_empty_log().into_log();
+        for (k, v) in self.fields.drain() {
+            if let Err(err) = v.insert_into(k, &mut event) {
+                warn!("failed to merge values for field: {}", err);
+            }
+        }
+        event
+    }
+}
+
+//------------------------------------------------------------------------------
+
+pub struct Transaction {
+    expire_after: Duration,
+    flush_period: Duration,
+    identifier_fields: Vec<Atom>,
+    merge_strategies: IndexMap<String, MergeStrategy>,
+    transaction_merge_states: HashMap<Discriminant, TransactionState>,
```

Comment:
I also note that since this is never cleared out and reset, it'll only ever grow in memory, so a 1 second spike of say 100k IDs would mean forever Vector reserved enough for at least 100k IDs, even if it never saw that much again.

Reply:
This is a good point but I don't know exactly when we should do the resize. Given that all of the meaningfully sized state is heap-allocated, I'm not too concerned about handling this right away.